src/Microsoft.DotNet.Tools.Builder/CompileContext.cs

```diff
@@ -52,22 +55,85 @@ public bool Compile(bool incremental)
             CreateOutputDirectories();
 
             //compile dependencies
-            foreach (var dependency in Sort(_dependencies))
+            foreach (var dependency in Sort(_dependencies.ProjectDependenciesWithSources))
             {
-                if (!InvokeCompileOnDependency(dependency, _outputPath, _intermediateOutputPath))
+                if (incremental)
+                {
+                    var dependencyContext = ProjectContext.Create(dependency.Path, dependency.TargetFrameworkInfo.FrameworkName);
+
+                    if (!NeedsRebuilding(dependencyContext, new ProjectDependenciesFacade(dependencyContext, _args.ConfigValue)))
+                    {
+                        continue;
+                    }
+                }
+
+                if (!InvokeCompileOnDependency(dependency))
                 {
                     return false;
                 }
             }
 
+            if (incremental && !NeedsRebuilding(_rootProject, _dependencies))
+            {
+                //todo: what if the previous build had errors / warnings and nothing changed? Need to propagate them in case of incremental
+                return true;
+            }
+
             //compile project
-            var success = InvokeCompileOnRootProject(_outputPath, _intermediateOutputPath);
+            var success = InvokeCompileOnRootProject();
 
             PrintSummary(success);
 
             return success;
         }
 
+        private bool NeedsRebuilding(ProjectContext project, ProjectDependenciesFacade dependencies)
+        {
+            var compilerIO = GetCompileIO(project, _args.ConfigValue, _args.OutputValue, _args.IntermediateValue, dependencies);
+
+            if (! (compilerIO.Outputs.Any() && compilerIO.Inputs.Any()))
```

Comment:
Couldn't you do all of this logic with a single for loop over inputs and outputs?